pallets/pallet-lottery/src/staking/mod.rs

```diff
@@ -0,0 +1,344 @@
+        }
+        log::debug!("Deposits: {:?}", deposits);
+        if !remaining_deposit.is_zero() {
+            log::error!(
+                "FATAL: We have {:?} unstaked balance left over after depositing, returning empty vec",
+                remaining_deposit
+            );
+            deposits.clear();
+        }
+        deposits
+    }
+
+    #[named]
+    pub(crate) fn calculate_withdrawal_distribution(
+        withdrawal_amount: BalanceOf<T>,
+    ) -> Vec<T::AccountId> {
+        log::trace!(function_name!());
+        if withdrawal_amount.is_zero() {
+            return vec![];
+        }
+        let mut withdrawals = vec![];
+        let mut remaining_balance = withdrawal_amount;
+
+        // Only unstake collators we're staked to and not already unstaking from
+        let staked_collators: Vec<_> = StakedCollators::<T>::iter_keys().collect();
+        let collators_we_are_unstaking_from: Vec<_> = UnstakingCollators::<T>::get()
+            .iter()
+            .cloned()
+            .map(|uc| uc.account)
+            .collect();
+        // NOTE: This is O(n^2) but both vecs are << 100 elements
+        let withdrawal_eligible_collators: Vec<_> = staked_collators
+            .iter()
+            .filter(|account| !collators_we_are_unstaking_from.contains(account))
+            .cloned()
+            .collect();
+        if withdrawal_eligible_collators.is_empty() {
+            return vec![];
+        }
+        // first concern: If there are inactive collators we are staked with, prefer these
+        let (mut collators, balance_unstaked) = withdraw_strategies::unstake_inactive_collators::<T>(
+            &withdrawal_eligible_collators,
+            remaining_balance,
+        );
+        withdrawals.append(&mut collators);
+        remaining_balance = remaining_balance.saturating_sub(balance_unstaked);
+        if remaining_balance.is_zero() && !withdrawals.is_empty() {
+            return withdrawals;
+        }
+        // If we have balance to withdraw left over, we have to unstake some healthy collator.
+        // Unstake starting from the highest overallocated collator ( since that yields the lowest APY ) going down until request is satisfied
+        let (mut collators, balance_unstaked) = withdraw_strategies::unstake_least_apy_collators::<T>(
+            &withdrawal_eligible_collators
+                .into_iter()
+                .filter(|collator| !withdrawals.contains(collator))
+                .collect(),
+            remaining_balance,
+        );
+        withdrawals.append(&mut collators);
+        remaining_balance = remaining_balance.saturating_sub(balance_unstaked);
+
+        if !remaining_balance.is_zero() {
+            log::error!(
+                "FATAL: We have {:?} left that COULD NOT BE UNSTAKED",
+                remaining_balance
+            );
+            return vec![]; // NOTE: Ensure we do not continue with the partial withdrawal
+        }
+        // TODO: Add random fallback
```

Comment:
is there an issue for this ?

Reply:
i'll remove this, that comment was a spur-of-the-moment thing and is not actually necessary